File: server/src/main/scala/producer/message_generator/data_generators/relative_date_time/RelativeDateTimeGenerator.scala
package producer.message_generator.data_generators.relative_date_time

import com.tools.teal.pulsar.ui.producer.v1.producer as pb
import producer.message_generator.json_generator.JsonGenerator

case class RelativeDateTimeGenerator(
    generator: FixedRelativeDateTimeGenerator | JsonGenerator
)

object RelativeDateTimeGenerator:
    def fromPb(v: pb.RelativeDateTimeGenerator): RelativeDateTimeGenerator =
        val generator = v.generator match
            case pb.RelativeDateTimeGenerator.Generator.GeneratorFixedRelativeDateTime(v) => FixedRelativeDateTimeGenerator.fromPb(v)
            case pb.RelativeDateTimeGenerator.Generator.GeneratorFromJsonNumberMs(v)      => JsonGenerator.fromPb(v)
            case _                                                                        => throw new Exception("Unknown relative date time generator type")

        RelativeDateTimeGenerator(
            generator = generator
        )

    def toPb(v: RelativeDateTimeGenerator): pb.RelativeDateTimeGenerator =
        val generator = v.generator match
            case v: FixedRelativeDateTimeGenerator =>
                pb.RelativeDateTimeGenerator.Generator.GeneratorFixedRelativeDateTime(FixedRelativeDateTimeGenerator.toPb(v))
            case v: JsonGenerator => pb.RelativeDateTimeGenerator.Generator.GeneratorFromJsonNumberMs(JsonGenerator.toPb(v))

        pb.RelativeDateTimeGenerator(
            generator = generator
        )

File: ui/components/ui/ProducerSession/producer-task/message-generator/data-generators/relative-date-time/relative-date-time-generator.ts
import * as pb from '../../../../../../../grpc-web/tools/teal/pulsar/ui/producer/v1/producer_pb';
import { JsonGenerator, jsonGeneratorFromPb, jsonGeneratorToPb } from '../json-generator/json-generator';
import { FixedRelativeDateTimeGenerator, fixedRelativeDateTimeGeneratorFromPb, fixedRelativeDateTimeGeneratorToPb } from './fixed-relative-date-time-generator';

export type RelativeDateTimeGenerator = {
  type: 'relative-date-time-generator',
  generator: FixedRelativeDateTimeGenerator | { type: 'relative-date-time-generator-from-json-number-millis', generator: JsonGenerator }
};

export function relativeDateTimeGeneratorFromPb(v: pb.RelativeDateTimeGenerator): RelativeDateTimeGenerator {
  switch (v.getGeneratorCase()) {
    case pb.RelativeDateTimeGenerator.GeneratorCase.GENERATOR_FIXED_RELATIVE_DATE_TIME:
      return {
        type: 'relative-date-time-generator',
        generator: fixedRelativeDateTimeGeneratorFromPb(v.getGeneratorFixedRelativeDateTime()!)
      };
    case pb.RelativeDateTimeGenerator.GeneratorCase.GENERATOR_FROM_JSON_NUMBER_MILLIS:
      return {
        type: 'relative-date-time-generator',
        generator: {
          type: 'relative-date-time-generator-from-json-number-millis',
          generator: jsonGeneratorFromPb(v.getGeneratorFromJsonNumberMillis()!)
        }
      };
    default:
      throw new Error(`Unknown generator case: ${v.getGeneratorCase()}`);
  }
}

export function relativeDateTimeGeneratorToPb(v: RelativeDateTimeGenerator): pb.RelativeDateTimeGenerator {
  const generatorPb = new pb.RelativeDateTimeGenerator();

  switch(v.generator.type) {
    case 'fixed-relative-date-time-generator': {
      generatorPb.setGeneratorFixedRelativeDateTime(fixedRelativeDateTimeGeneratorToPb(v.generator));
      break;
    }
    case 'relative-date-time-generator-from-json-number-millis': {
      generatorPb.setGeneratorFromJsonNumberMillis(jsonGeneratorToPb(v.generator.generator));
      break;
    }
  }

  return generatorPb;
}
